src/extensions/idle.rs

```diff
@@ -100,37 +130,74 @@ impl<'a, T: Read + Write + 'a> Handle<'a, T> {
+            let rest = match self.session.readline(&mut v) {
                 Err(Error::Io(ref e))
                     if e.kind() == io::ErrorKind::TimedOut
                         || e.kind() == io::ErrorKind::WouldBlock =>
                 {
-                    if reconnect {
-                        self.terminate()?;
-                        self.init()?;
-                        return self.wait_inner(reconnect);
+                    break Ok(WaitOutcome::TimedOut);
+                }
+                Ok(_len) => {
+                    //  Handle Dovecot's imap_idle_notify_interval message
+                    if v.eq_ignore_ascii_case(b"* OK Still here\r\n") {
+                        v.clear();
+                        continue;
+                    }
+                    match parse_idle(&v) {
+                        (_rest, Some(Err(r))) => break Err(r),
+                        (rest, Some(Ok(response))) => {
+                            if !callback(response) {
+                                break Ok(WaitOutcome::MailboxChanged);
+                            }
+                            rest
+                        }
+                        (rest, None) => rest,
                     }
-                    Ok(WaitOutcome::TimedOut)
                 }
-                Ok(()) => Ok(WaitOutcome::MailboxChanged),
-                Err(r) => Err(r),
-            }?;
+                Err(r) => break Err(r),
+            };
 
-            // Handle Dovecot's imap_idle_notify_interval message
-            if v.eq_ignore_ascii_case(b"* OK Still here\r\n") {
+            // Update remaining data with unparsed data if needed.
+            if rest.is_empty() {
                 v.clear();
             } else {
-                break Ok(result);
+                // Assert on partial parse in debug builds - we expect to always parse all
+                // or none of the input buffer. On release builds, we still do the right thing.
+                debug_assert!(
+                    rest.len() != v.len(),
+                    "Unexpected partial parse: input: {:?}, output: {:?}",
+                    v,
+                    rest
+                );
+                let used = v.len() - rest.len();
+                v.drain(0..used);
+            }
+        };
+
+        // Reconnect on timeout if needed
+        match (reconnect, result) {
+            (true, Ok(WaitOutcome::TimedOut)) => {
+                self.terminate()?;
+                self.init()?;
+                self.wait_inner(reconnect, callback)
             }
+            (_, result) => result,
         }
     }
 
-    /// Block until the selected mailbox changes.
-    pub fn wait(mut self) -> Result<()> {
-        self.wait_inner(true).map(|_| ())
+    /// Block until the given callback returns `false`, or until a response
+    /// arrives that is not explicitly handled by [`UnsolicitedResponse`].
+    pub fn wait<F>(mut self, callback: F) -> Result<()>
```

Comment:
I think a good compromise here might be to name the function `wait_while` (and similarly for the other methods). That way it _should_ be clear what the `true`/`false` actually means. What do you think?

Reply:
Oh, I am not super torqued about it? I feel like either way you'll have to read the docs to know for sure. And I like the simplicity of `wait` :)

I was also thinking we could just provide a `pub const CONTINUE: bool = true`, and `pub const STOP: bool = false`, but this feels slightly silly? It's super normal in C/C++, but doesn't seem super normal in Rust?

Reply:
I think I prefer `wait_while` since it also leans on the precedent set in the standard library with things like `take_while`. While `wait` is nice and short, it's not _super_ descriptive of what the function now actually does, whereas I think `wait_while` is.

So, I've seen that occasionally in Rust, but only defined in the _user_ of the library, rarely in the library itself. As in, `imap` wouldn't define those `const`s, but rather the user of `imap` would define it to make their calls clearer. But I think with a rename to `wait_while`, that's already clear enough!

Reply:
Done. I went with the pattern `wait_thing_while()` when there were more words in the function signature, instead of `wait_while_thing`, since the latter seems misleading (ie. `wait_while_keepalive` is confusing).

I also only did the public functions, since I don't think the internal ones don't need the usage hint?

Since we are changing the API it might be nice to switch this to a builder pattern so there is only one wait, ie.:

```
session.idle()
    .keepalive()
    .wait_while(|res| {
        // do something
    })
```

But this can happen later if at all .

Reply:
Thank you! I agree builder would be nice here. Happy to do in this PR or in a follow-up — up to you.

Reply:
Let's do a follow up, since it'll be fine as a standalone and we the TLS one to do as well.